### PR TITLE
Support system properties as a way of providing configuration options to the CLI

### PR DESCRIPTION
**Pull request by jccampanero**
## Impact

- [ ] Bug fix (non-breaking change which fixes expected existing functionality)
- [x] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Description

The `LiquibaseLauncher` code, and the Liquibase CLI in general, have been enriched to support passing the different configuration options as JVM system properties.

Previously, such information could only be passed as operating system environment variables.

Fixes and augments, in order to provide a uniform way of reading the different configuration options, the scope of #3777.

## Things to be aware of

Nothing.

## Things to worry about

Nothing.

## Additional Context

No test have been created for the issue.

### CODE REVIEW

File: liquibase-cli/src/main/java/liquibase/integration/commandline/LiquibaseLauncher.java [hunk 1]
```diff
@@ -122,4 +131,13 @@ private static void debug(String message, Throwable e) {
             }
         }
     }
+
+    private static String getValueFromJvmPropertyOrEnv(String name) {
+        String value = System.getProperty(name);
```
Comment (nvoxland) [inline liquibase-cli/src/main/java/liquibase/integration/commandline/LiquibaseLauncher.java:136]: Generally system properties are more in the format `liquibase.home` or `liquibase.launcher.debug`

Can you translate the env varible format to the other and check for it as a system property too?

Comment (jccampanero) [reply liquibase-cli/src/main/java/liquibase/integration/commandline/LiquibaseLauncher.java:136]: It makes perfect sense to me, I agree with you.

Of course, I will implement the changes tonight.

Comment (jccampanero) [reply liquibase-cli/src/main/java/liquibase/integration/commandline/LiquibaseLauncher.java]: I updated the code to provide a more appropriate naming convention for the JVM properties.

Please, don't hesitate to contact me if further changes should be performed.

**Comment by nvoxland** [comment]
**Initial/Pre-Review Thoughts**

The changes look good to me. Thanks!

Questions I have:
- None

Potential risks:
- None

What could make the full review difficult:
- Nothing

**Review by nvoxland — APPROVED** [review]

**Comment by jccampanero** [comment]
You are welcome @nvoxland Nathan.

**Comment by filipelautert** [comment]
Hello @jccampanero ,maybe instead of using a Map from string to string, could not we use an Enum class where they key is similar to the string name (LIQUIBASE_HOME_JVM_PROPERTY_NAME ) and it would have a "value" field that would be the parameter value ("liquibase.home") . What do you think? This way we don't need to populate maps, etc.

**Comment by jccampanero** [comment]
> Hello @jccampanero ,maybe instead of using a Map from string to string, could not we use an Enum class where they key is similar to the string name (LIQUIBASE_HOME_JVM_PROPERTY_NAME ) and it would have a "value" field that would be the parameter value ("liquibase.home") . What do you think? This way we don't need to populate maps, etc.

Sorry for the late reply @filipelautert.

Yes, of course, I updated the code as requested. I used `jvmPropertyName` instead of `value` in the `enum` definition because I think it is more precise but please, feel free to adapt the code if you wish.

**Comment by filipelautert** [comment]
> Yes, of course, I updated the code as requested. I used `jvmPropertyName` instead of `value` in the `enum` definition because I think it is more precise but please, feel free to adapt the code if you wish.

Nice, I believe it is better now as we can see that those values are used only for this purpose, thanks for applying the changes!

**Comment by jccampanero** [comment]
You are welcome @filipelautert, my pleasure.

**Review by filipelautert — APPROVED** [review]

**Comment by rberezen** [comment]
Hello, @jccampanero! I attempted to execute these two commands `java -Dliquibase.home='/home/rberezenskyi/Downloads/liquibase' -jar "./internal/lib/liquibase-core.jar" --version` and `JAVA_OPTS="-Dliquibase.home=/home/rberezenskyi/Downloads/liquibases" ./liquibase --version` to change the Liquibase Home directory, but unfortunately, I observed that the Liquibase Home setting did not change. Please refer to the attached screenshots for more information. Perhaps I'm making a mistake. Would you mind examining it?

Additionally, if the liquibase.home is set to a non-existent directory, an exception is thrown, as shown below:
```
Exception in thread "main" java.lang.ClassNotFoundException: liquibase.integration.commandline.LiquibaseCommandLine
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:476)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:589)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at liquibase.integration.commandline.LiquibaseLauncher.main(LiquibaseLauncher.java:172)
```
It would be greatly appreciated if a more user-friendly error message could be displayed in such cases. Thank you!
Screenshot from 2023-06-28 11-37-11
Screenshot from 2023-06-28 10-53-56

**Comment by jccampanero** [comment]
Of course @rberezen, I will look into it.

Thank you for reporting the issue.

**Comment by wwillard7800** [comment]
The `LIQUIBASE_HOME` directory is also referenced in in `LiquibaseCommandLine.java` in this way at line 1223

```
String liquibaseHome;
Path liquibaseHomePath = null;
try {
    liquibaseHomePath = new File(ObjectUtil.defaultIfNull(System.getenv("LIQUIBASE_HOME"), 
                   workingDirectory.toAbsolutePath().toString())).getAbsoluteFile().getCanonicalFile().toPath();
    liquibaseHome = liquibaseHomePath.toString();
 } catch (IOException e) {
     liquibaseHome = "Cannot resolve LIQUIBASE_HOME: " + e.getMessage();
 }

```
This code will need to be modified in order to utilize a system property.

**Comment by jccampanero** [comment]
Yes @wwillard7800, thank you very much, I found the same code fragment, the environment variable is referenced there too. 

If would be great if the already read setting could be passed as an argument to `LiquibaseCommandLine` from `LiquibaseLauncher` but the only referenced method from that class (as it should) is `main`.

Probably the best way to go will be repeat the behavior already implemented in `LiquibaseLauncher`, perform the same check looking for the System property first.

Thank you very much again Wesley!

**Comment by jccampanero** [comment]
I submitted a possible implementation of the requested changes @rberezen.

Please, don't hesitate to contact me again if you think that any type of improvement or additional work should be done.

**Comment by rberezen** [comment]
> I submitted a possible implementation of the requested changes @rberezen.
> 
> Please, don't hesitate to contact me again if you think that any type of improvement or additional work should be done.

Thank you @jccampanero for the update! The issue of encountering an unfriendly exception when the liquibase.home is set to a non-existent directory has been resolved! However, the Liquibase Home setting is still not updated if it is provided with JAVA_OPTS. Tried the same commands: `commands java -Dliquibase.home='/home/rberezenskyi/Downloads/liquibase' -jar "./internal/lib/liquibase-core.jar" --version and JAVA_OPTS="-Dliquibase.home=/home/rberezenskyi/Downloads/liquibases" ./liquibase --version`

**Comment by jccampanero** [comment]
Sorry for the late reply @rberezen.

Please, let me check it out again. I will get back to you soon.

**Comment by jccampanero** [comment]
@rberezen Sorry for the late reply.

Please, could you double check your setup?

I tested the current code and although it looks erroneous at first glance for me too in the end the shade plugin and the generation of the `liquibase-core.jar` arfifact were the reasons of the problem, they weren't updating the cli classes properly. But, after updating the jar everything worked fine.

Please, could you try again?

**Comment by rberezen** [comment]
@jccampanero There seems to be an issue with my setup. Everything is functioning as expected. Apologies for any confusion caused.

**Comment by rberezen** [comment]
This pull request introduces the capability to set the following settings as JAVA_OPTS environment variables: liquibase.home, liquibase.launcher.debug, and liquibase.launcher.parent_classloader.

- Functional and test harness executions passing.
- Manual testing completed.

APPROVED

**Review by rberezen — APPROVED** [review]

**Comment by jccampanero** [comment]
> @jccampanero There seems to be an issue with my setup. Everything is functioning as expected. Apologies for any confusion caused.

There is no need to apologize @rberezen; on the contrary, thank you very much for your support.

**Comment by filipelautert** [comment]
@rberezen do you think we need a PD ticket to document this new option?

**Comment by rberezen** [comment]
> @rberezen do you think we need a PD ticket to document this new option?

Thanks, for the reminder. Created [PD-3610](https://datical.atlassian.net/browse/PD-3610)

**Comment by sayaliM0412** [comment]
Hello @jccampanero, I am testing some internal build process and using your branch as a guinea pig. Please let me know if you face any issues during this process. Thank you.

cc: @filipelautert

**Comment by jccampanero** [comment]
> Hello @jccampanero, I am testing some internal build process and using your branch as a guinea pig. Please let me know if you face any issues during this process. Thank you.
> 
> cc: @filipelautert

It is perfectly fine @sayaliM0412, please, feel free to use the branch for your tests, as you consider it appropriate.

**Comment by adrian-velonis1** [comment]
Internal: https://datical.atlassian.net/browse/PD-3610

**Pull request merged**